text/0000-disjoint-groups/0000-disjoint-groups.md

```diff
@@ -0,0 +1,149 @@
+
+As mentioned above, the current group message routing mechanism is not secure. To prevent intercepting group messages, we will instead relay them from group to group, verifying quorum in every step:
+
+Assume a node `n` in group `B` receives a group message signed by someone in group `A`:
+
+* If `A` is not connected to `B`, drop the message.
+* If this message has not yet been signed by a quorum of members of `A`, keep it in the cache.
+* If it has been signed by a quorum of members of `A`: If we are the recipient, handle it. If not, find the connected group `C` closest to the destination and send the message, signed by `n`, to _every member_ of `C`.
+
+This creates a number of direct messages in each hop that grows quadratically in the group size. To reduce network traffic, the `route`-th node of group `B` could accumulate `A`'s signatures instead, making the message number linear in the group size, at the cost of having effectively two network hops per group hop:
+
+* If `A` is not connected to `B`, drop the message.
+* If this message has not yet been signed by a quorum of members of `A`, keep it in the cache.
+* If it has been signed by a quorum of members of `A`: If we are the recipient, handle it. If not, find the connected group `C` closest to the destination and send the message, signed by `n`, to the `route`-th closest member of `C`.
+* If it has been signed by a quorum of members of `A` and we are the `route`-th closest member of `B`, send the message with `A`'s signatures to every member of `B`.
+
+The quorum cannot be a constant anymore, due to varying group sizes. It needs to be a percentage strictly greater than 50% instead, and in a group of size `n`, a number `x` of nodes will constitute a quorum if `x / n >= QUORUM`.
+
+### Joining nodes
+
+A new node needs to first connect to everyone in its target group. After that, it can drop its proxy node. The other group members can provide the node with their routing tables, which should all be identical. It can then make connections to these other groups, too.
+
+### Group split
+
+If a group `(p)` splits, all its nodes stay connected: The new groups are now each other's `i`-th bucket, where `i` is the number of bits in `p`.
+
+The group sends a direct message to all its contacts to inform them about the split. These need to note that the group has split, even if they stay connected to all its members.
+
+The node can then disconnect immediately (no more `ConnectionUnneeded` required) from every group `(q)` for which `q` now differs in more than one bit from its new group prefix.
```

Comment:
I think this then needs a Splitting message to vaults, same for converge perhaps. If vaults do have DataChains then aany data on converge again needs to republish (i.e. send all datablocks from the previous group which was CLB (Common Leading Bits) + 1 == our address).